test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java

```diff
@@ -2492,7 +2448,7 @@ public void ensureEstimatedStats() {
     }
 
     @Override
-    public void assertAfterTest() throws IOException {
+    public synchronized void assertAfterTest() throws IOException {
```

Comment:
This method accesses the cluster and holding the monitor while doing so could potentially lead to deadlocks?

Reply:
I think this should never run in parallel with other operations modifying the cluster and should not modify the cluster itself => Since we should only do reads on the state of cluster I think we won't dead-lock here ever and it's a lot easier to reason about a series of asserts when you at least have the guarantee that the cluster doesn't change between them isn't it?